### PR TITLE
gh-120057: Add os.environ.refresh() method

### PR DESCRIPTION
**Pull request by vstinner**
* Issue: gh-120057

----
📚 Documentation preview 📚: https://cpython-previews--120059.org.readthedocs.build/

### CODE REVIEW

File: Lib/os.py [hunk 1]
```diff
@@ -773,7 +781,18 @@ def __ror__(self, other):
...
+    if _exists("_create_environ"):
+        def refresh(self):
+            environ = _create_environ()
+            if name == 'nt':
+                data = {}
+                for key, value in environ.items():
+                    data[encodekey(key)] = value
+            else:
+                data = environ
+            self._data = data
```
Comment (eryksun) [inline Lib/os.py]: If `os.environb` exists, then `os.environ` and `os.environb` will be out of sync after calling `refresh()` if it's implemented by reassigning `_data`. Maybe this should instead `clear()` and `update()`.

```suggestion
            self._data.clear()
            self._data.update(data)
```

File: Doc/library/os.rst [hunk 2]
```diff
@@ -193,6 +193,9 @@ process and user.
    to the environment made after this time are not reflected in :data:`os.environ`,
    except for changes made by modifying :data:`os.environ` directly.
 
+   The :meth:`os.environ.refresh() <os.environ>` method updates
```
Comment (JelleZijlstra) [inline Doc/library/os.rst]:
```suggestion
   The :meth:`!os.environ.refresh()` method updates
```
It doesn't seem helpful to link back to `os.environ` here

File: Lib/os.py (hunk 1 above)
Comment (vstinner) [reply Lib/os.py]: Oh right, I added a test and used your code, thanks!

**Comment by vstinner** [comment]
I rebased the PR to fix a merge conflict.

I addressed @eryksun's and @JelleZijlstra's reviews: thanks!

File: Doc/library/os.rst [hunk 3]
```diff
@@ -222,6 +226,9 @@ process and user.
    :data:`os.environ`, and when one of the :meth:`pop` or :meth:`clear` methods is
    called.
 
+   .. versionchanged:: 3.14
```
Comment (picnixz) [inline Doc/library/os.rst]: Maybe put that one below `versionchanged:: 3.9`?

Comment (vstinner) [reply Doc/library/os.rst]: done

**Comment by vstinner** [comment]
@serhiy-storchaka @erlend-aasland: Would you mind to review this change?

File: Lib/os.py [hunk 4]
```diff
@@ -773,7 +781,21 @@ def __ror__(self, other):
...
         return new
 
-def _createenviron():
+    if _exists("_create_environ"):
+        def refresh(self):
+            environ = _create_environ()
+            if name == 'nt':
+                data = {}
+                for key, value in environ.items():
+                    data[self.encodekey(key)] = value
```
Comment (serhiy-storchaka) [inline Lib/os.py]: You can use a dict comprehension here.

**Review by serhiy-storchaka — COMMENTED** [review]
What about `os.environb.refresh()`? Does it exist? Does it work?

File: Lib/os.py (hunk 4 above)
Comment (vstinner) [reply Lib/os.py]: Done

**Comment by vstinner** [comment]
@serhiy-storchaka: I addressed your review. Would you mind to review my updated PR?

> What about os.environb.refresh()? Does it exist? Does it work?

I added a test on os.environb.refresh(). Yes, it works as expected :-)

**Review by serhiy-storchaka — COMMENTED** [review]
Should `os.environb.refresh()` be documented? We perhaps do not want to attract too much attention to it, as its effect is the same as of `os.environ.refresh()`.

File: Doc/library/os.rst [hunk 5]
```diff
@@ -193,6 +193,10 @@ process and user.
    to the environment made after this time are not reflected in :data:`os.environ`,
    except for changes made by modifying :data:`os.environ` directly.
 
+   The :meth:`!os.environ.refresh()` method updates
+   :data:`os.environ` with changes to the environment made in the same process
+   outside Python.
```
Comment (eryksun) [inline Doc/library/os.rst]:
```suggestion
   outside Python or by :func:`os.putenv`.
```

Comment (eryksun) [reply Doc/library/os.rst]: Alternatively, `os.putenv()` could be wrapped to update `os.environ`.

Comment (vstinner) [reply Doc/library/os.rst]: I applied your suggestion.

**Comment by vstinner** [comment]
> Should os.environb.refresh() be documented? We perhaps do not want to attract too much attention to it, as its effect is the same as of os.environ.refresh().

Hum, I prefer to not document it at https://docs.python.org/dev/library/os.html#os.environb :-)

**Comment by vstinner** [comment]
(Ooops, I closed on [Close] instead of [Comment], sorry.)

File: Doc/library/os.rst [hunk 6]
```diff
@@ -195,7 +195,7 @@ process and user.
 
    The :meth:`!os.environ.refresh()` method updates
    :data:`os.environ` with changes to the environment made in the same process
-   outside Python.
+   outside Python or by :func:`os.putenv`.
```
Comment (serhiy-storchaka) [inline Doc/library/os.rst]: Or `os.unsetenv`?

File: Lib/test/test_os.py [hunk 7]
```diff
@@ -1301,26 +1301,25 @@ def test_ror_operator(self):
...
-            from posix import putenv
-        except ImportError:
-            from nt import putenv
+        # Test os.environ.refresh() with putenv() which doesn't update
+        # os.environ
+        has_environb = hasattr(os, 'environb')
 
         os.environ['test_env'] = 'python_value'
-        putenv("test_env", "new_value")
+        os.putenv("test_env", "new_value")
```
Comment (serhiy-storchaka) [inline Lib/test/test_os.py:1307]: Please add `os.unsetenv` in the test. Note that the test will still be passed if remove `self._data.clear()`.

File: Lib/os.py [hunk 8]
```diff
@@ -773,7 +781,20 @@ def __ror__(self, other):
         new.update(self)
         return new
 
-def _createenviron():
+    if _exists("_create_environ"):
+        def refresh(self):
+            environ = _create_environ()
+            if name == 'nt':
+                data = {self.encodekey(key): value
```
Comment (serhiy-storchaka) [inline Lib/os.py:788]: You can rebind the same variable and remove the `else` branch.

File: Lib/test/test_os.py [hunk 9]
```diff
@@ -1298,6 +1298,35 @@ def test_ror_operator(self):
         self._test_underlying_process_env('_A_', '')
         self._test_underlying_process_env(overridden_key, original_value)
 
+    @unittest.skipUnless(hasattr(os.environ, 'refresh'),
```
Comment (serhiy-storchaka) [inline Lib/test/test_os.py]: How can it be absent?

**Comment by vstinner** [comment]
@serhiy-storchaka: PR updated for os.unsetenv().

Oh, I didn't know that Python has two functions (putenv, unsetenv) which modify directly the environment without updating `os.environ`.

File: Lib/test/test_os.py [hunk 10]
```diff
@@ -1316,6 +1314,17 @@ def test_refresh(self):
         if has_environb:
             self.assertEqual(os.environb[b'test_env'], b'new_value')
 
+        # Test with unsetenv() which doesn't update os.environ
```
Comment (serhiy-storchaka) [inline Lib/test/test_os.py:1317]: Maybe add this also for `os.environb.refresh()` for symmetry?

**Review by serhiy-storchaka — APPROVED** [review]
LGTM.

**Review by serhiy-storchaka — APPROVED** [review]

**Review by serhiy-storchaka — APPROVED** [review]

File: Doc/library/os.rst [hunk 11]
```diff
@@ -193,6 +193,10 @@ process and user.
    to the environment made after this time are not reflected in :data:`os.environ`,
    except for changes made by modifying :data:`os.environ` directly.
 
+   The :meth:`!os.environ.refresh()` method updates :data:`os.environ` with
```
Comment (erlend-aasland) [inline Doc/library/os.rst:196]: FTR, the `:meth:` and `:func:` roles will implicitly add parentheses in the rendered output; you do not need to add them explicitly.

**Pull request merged**